storybookjs/storybook
# Core: Story context is prepared before for supporting fine grained updates

Issue: #20584 



## What I did

I needed to make some changes to how we prepare stories, in order to make args reactive in frameworks that use reactive proxies such as SolidJS or Vue for fine-grained updates.

This change modifies the way Storybook currently prepares stories which causes reactive args to get lost for fine-grained updates JS frameworks. That's because for those frameworks handle args/props as proxies behind the scenes to make reactivity work. So when we do the `argType` mapping in `prepareStory` the Proxies are destroyed and args becomes a plain object again, losing the reactivity.

The proposal here is to change the flow, passing the mapped args at `renderToCanvas` so that the proxies stay intact. 

There's a chance that this change will also make reactive args in Vue3 possible, something that we currently can't do without re-creating the whole Vue story.

> **Note**
> This change means that args will be mapped _prior_ to being passed to decorators, where previously decorators received raw args, they now receive mapped args. This is the breaking change.


## How to test

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

- Unit tests:
```
prepareStory.test.ts
StoryRender.test.ts
```
@tmeasday - It was added a test at `prepareStory.test.ts` to confirm that prepared args are received on each decorator.

## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

      make sure to add the `ci:merged` or `ci:daily` GH label to it.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
